### ISSUE TITLE
Java architecture is too blackbox

### ISSUE THREAD

**Issue by amihaiemil** [yegor256/jpeek#96]
As I explained in [this](https://github.com/yegor256/jpeek/pull/95#issuecomment-357950881) comment, the Java architecture tightens the reports too much. 

When a new report is created, a new metric implemented, there is an XML/XSLT file added which represents said metric and that file is used by class Report (which has no interface and is very configurable, btw). Then, the App class iterates over all instances of Report and asks for the generated report... 

This is a big drawback for testing. How do you test a single report? All that is tested so far (as far as I can tell) is the format of the final output, but not the correctness of it (I'm talking about numbers). 

First of all, I would represent each report in its own class which should be both unit tested and integration tested. Of course, implementing in XSLT is fine, I am referring strictly ot Java here. 

This ticket should be broken in multiple ones, for sure, but let's discuss first...

**Comment by 0crat** [comment]
@yegor256[/z](http://www.0crat.com/u/yegor256) please, pay attention to this issue

**Comment by yegor256** [comment]
@amihaiemil have you seen `MetricsTest`?

**Comment by amihaiemil** [comment]
@yegor256 Yes, and to be honest, I couldn't make much of it. Looking closer now, I guess it tests the numeric output (matches agains this.target, this.value). 

But I still think each metric should be an implementation of Report and be unit tested by itself too, right?

**Comment by yegor256** [comment]
@amihaiemil well, each metric is just a small XSLT script, which produces some output. The only thing we are concerned about is the numbers. What else to test here?

**Comment by amihaiemil** [comment]
@yegor256 I see... in fact, I wanted to implement those metrics in Java (with *some* help from XSLT to leverage it), in their own classes. Apparently that was the initial architecture and the decision was made to go 100% with XSLT. 

I close this, then, thanks :)

**Comment by 0crat** [comment]
Job `gh:yegor256/jpeek#96` is not assigned, can't get start time

**Comment by 0crat** [comment]
Job [#96](https://github.com/yegor256/jpeek/issues/96) was not in scope

**Comment by yegor256** [comment]
@amihaiemil yes, XSLT is way more expressive and it's much better to parse Java only once and then deal with XMLs.
